src/ImageSharp/IO/ChunkedMemoryStream.cs

```diff
@@ -94,247 +57,133 @@ public override long Position
         => throw new NotSupportedException();
 
     /// <inheritdoc/>
-    protected override void Dispose(bool disposing)
+    public override int ReadByte()
     {
-        if (this.isDisposed)
-        {
-            return;
-        }
-
-        try
-        {
-            this.isDisposed = true;
-            if (disposing)
-            {
-                ReleaseMemoryChunks(this.memoryChunk);
-            }
-
-            this.memoryChunk = null;
-            this.writeChunk = null;
-            this.readChunk = null;
-            this.chunkCount = 0;
-        }
-        finally
+        this.EnsureNotDisposed();
+        if (this.position >= this.length)
         {
-            base.Dispose(disposing);
+            return -1;
         }
-    }
 
-    /// <inheritdoc/>
-    public override void Flush()
-    {
+        _ = this.Read(this.singleByteBuffer, 0, 1);
+        return MemoryMarshal.GetReference<byte>(this.singleByteBuffer);
     }
 
     /// <inheritdoc/>
-    [MethodImpl(MethodImplOptions.AggressiveInlining)]
     public override int Read(byte[] buffer, int offset, int count)
     {
+        this.EnsureNotDisposed();
+
         Guard.NotNull(buffer, nameof(buffer));
         Guard.MustBeGreaterThanOrEqualTo(offset, 0, nameof(offset));
         Guard.MustBeGreaterThanOrEqualTo(count, 0, nameof(count));
 
         const string bufferMessage = "Offset subtracted from the buffer length is less than count.";
         Guard.IsFalse(buffer.Length - offset < count, nameof(buffer), bufferMessage);
 
-        return this.ReadImpl(buffer.AsSpan(offset, count));
+        return this.Read(buffer.AsSpan(offset, count));
     }
 
     /// <inheritdoc/>
-    [MethodImpl(MethodImplOptions.AggressiveInlining)]
-    public override int Read(Span<byte> buffer) => this.ReadImpl(buffer);
-
-    private int ReadImpl(Span<byte> buffer)
+    public override int Read(Span<byte> buffer)
     {
         this.EnsureNotDisposed();
 
-        if (this.readChunk is null)
-        {
-            if (this.memoryChunk is null)
-            {
-                return 0;
-            }
+        int offset = 0;
+        int count = buffer.Length;
 
-            this.readChunk = this.memoryChunk;
-            this.readOffset = 0;
+        long remaining = this.length - this.position;
+        if (remaining <= 0)
+        {
+            // Already at the end of the stream, nothing to read
+            return 0;
         }
 
-        IMemoryOwner<byte> chunkBuffer = this.readChunk.Buffer;
-        int chunkSize = this.readChunk.Length;
-        if (this.readChunk.Next is null)
+        if (remaining > count)
         {
-            chunkSize = this.writeOffset;
+            remaining = count;
         }
 
+        int bytesToRead = (int)remaining;
         int bytesRead = 0;
-        int offset = 0;
-        int count = buffer.Length;
-        while (count > 0)
+        while (bytesToRead != 0 && this.currentChunk != this.memoryChunkBuffer.Length)
```

Comment:
Even though current code looks good and `Slice` would throw for a negative number, I find this safer to maintain. Same for `Write`.

```suggestion
        while (bytesToRead >= 0 && this.currentChunk != this.memoryChunkBuffer.Length)
```